google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/GradleBuildDeploymentSource.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2016 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.cloud;
+
+import com.intellij.openapi.components.ServiceManager;
+import com.intellij.openapi.externalSystem.service.project.manage.ProjectDataManager;
+import com.intellij.openapi.module.ModulePointer;
+import com.intellij.openapi.project.Project;
+import com.intellij.remoteServer.configuration.deployment.DeploymentSourceType;
+import com.intellij.remoteServer.impl.configuration.deployment.ModuleDeploymentSourceImpl;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.plugins.gradle.model.ExternalProject;
+import org.jetbrains.plugins.gradle.service.project.data.ExternalProjectDataService;
+import org.jetbrains.plugins.gradle.util.GradleConstants;
+
+import java.io.File;
+import java.io.FilenameFilter;
+
+import javax.swing.Icon;
+
+import icons.GradleIcons;
+
+/**
+ * A deployment source backed by the Gradle build system
+ */
+public class GradleBuildDeploymentSource extends ModuleDeploymentSourceImpl {
+
+  private final Project project;
+
+  public GradleBuildDeploymentSource(@NotNull ModulePointer pointer, @NotNull Project project) {
+    super(pointer);
+    this.project = project;
+  }
+
+  @NotNull
+  @Override
+  public String getPresentableName() {
+    return String.format("Gradle build: %s", getModulePointer().getModuleName());
+  }
+
+  @Nullable
+  @Override
+  public Icon getIcon() {
+    return GradleIcons.Gradle;
+  }
+
+  @NotNull
+  @Override
+  public DeploymentSourceType<?> getType() {
+    return DeploymentSourceType.EP_NAME.findExtension(GradleBuildDeploymentSourceType.class);
+  }
+
+  @Nullable
+  @Override
+  public File getFile() {
```

Comment:
For a multi-module maven/gradle project, each module will have its own maven/gradle deployment source:

![image](https://cloud.githubusercontent.com/assets/1735744/14652388/1878e950-0642-11e6-8456-d947fd4a6e9b.png)

However, currently switching between them doesn't yet properly update the pre-deploy build task. TBD: how do we deal with ear producing modules, such as the one shown in the above screenshot, since these cannot be deployed to flex.